### DISCUSSION TITLE
How to restrict some string values

### DISCUSSION THREAD

**Discussion by piotrkulpinski** [colinhacks/zod#2794] [Q&A] [answered]
Hi, 

I'm trying to validate a string in Zod agains an array of reserved values, like this:

```ts
const reservedSlugs = ['app', 'admin', 'www']
```

Is there any easy way to validate that? I know I could possibly use regex, but I'm feeling like there should be an easier way to do that in Zod.

Thanks!

**Comment by mtsknn** [comment, marked as answer]
[Zod enums](https://zod.dev/?id=zod-enums)

```ts
const reservedSlugs = ['app', 'admin', 'www'] as const
const SlugSchema = z.enum(reservedSlugs)
```

**Reply by piotrkulpinski** [reply]
Thanks, but this will only allow the reserved slugs to be passed. I want the exact opposite. Anything **but** the reserved slugs should be allowed.

**Reply by mtsknn** [reply]
Whoops. :-) How about using [`.refine`](https://zod.dev/?id=refine)?

```ts
const SlugSchema = z.string().refine((slug) => !reservedSlugs.includes(slug), {
  message: "Slug can't be one of reserved slugs",
})
```

**Reply by piotrkulpinski** [reply]
Tried that, but that doesn't seems to be working. Also tried the refine method on the whole object (containing the slug) but without luck.

Here's my current schema: 
```ts
export const siteSchema = z.object({
	name: z.string().trim().min(3).max(32),
	slug: z.string().refine((slug) => !reservedSlugs.includes(slug), {
		message: "Slug can't be one of reserved slugs",
	}),
})
```

**Reply by mtsknn** [reply]
Using your schema, `siteSchema.parse({ name: 'foo', slug: 'bar' })` is valid, `siteSchema.parse({ name: 'foo', slug: 'www' })` throws an error.

Isn't that correct? Or what are you specifically trying to do?

**Reply by piotrkulpinski** [reply]
My bad. Turns out I was using the schema wrong!
Thanks a lot for your help!
